src/System.Private.CoreLib/shared/System/Activator.RuntimeType.cs

```diff
@@ -132,26 +132,135 @@ internal static object CreateInstance(Type type, bool nonPublic, bool wrapExcept
+            }
+
+            return instance;
+        }
 
-            // Skip the CreateInstanceCheckThis call to avoid perf cost and to maintain compatibility with V2 (throwing the same exceptions).
-            return (T)rt.CreateInstanceDefaultCtor(publicOnly: true, skipCheckThis: true, fillCache: true, wrapExceptions: true);
+        private delegate T InstanceFactory<T>(bool publicOnly, bool wrapExceptions);
+
+        private delegate void ConstructorOfClass<T>(T instance);
+        private delegate void ConstructorOfStruct<T>(ref T instance);
+
+        private static T CreateDelegate<T>(MethodBase methodInfo)
+            where T : Delegate
+        {
+            return (T)typeof(T)
+               .GetConstructor(new[] { typeof(object), typeof(IntPtr) })
+               .Invoke(new object[] { null, methodInfo.MethodHandle.GetFunctionPointer() });
+        }
+
+        private static class ActivatorCache
+        {
+            internal static readonly MethodInfo FactoryOfClassMethod = new InstanceFactory<object>(CreateInstanceOfClass<object>).Method.GetGenericMethodDefinition();
+            internal static readonly MethodInfo FactoryOfStructMethod = new InstanceFactory<int>(CreateInstanceOfStruct<int>).Method.GetGenericMethodDefinition();
+        }
+
+        private static class ActivatorCache<T>
+        {
+            internal static readonly RuntimeType Type;
+            internal static readonly bool IsValueType;
+            internal static readonly bool HasElementType;
+            internal static readonly bool HasNonPublicConstructor;
+
+            internal static readonly ConstructorOfClass<T> Construct;
+            internal static readonly ConstructorOfStruct<T> ConstructByRef;
+
+            internal static readonly InstanceFactory<T> Factory;
+
+            static ActivatorCache()
+            {
+                Type = (RuntimeType)typeof(T);
+                IsValueType = Type.IsValueType;
+                HasElementType = Type.HasElementType;
+
+                if (HasElementType)
+                {
+                    ConstructorInfo runtimeCtorInfo = Type.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, new[] { Type }, null);
+
+                    if (runtimeCtorInfo != null)
+                    {
+                        HasNonPublicConstructor = !runtimeCtorInfo.IsPublic;
+
+                        if (IsValueType)
+                            ConstructByRef = CreateDelegate<ConstructorOfStruct<T>>(runtimeCtorInfo);
+                        else
+                            Construct = CreateDelegate<ConstructorOfClass<T>>(runtimeCtorInfo);
+                    }
+                }
+
+                MethodInfo factoryMethod = IsValueType
+                    ? ActivatorCache.FactoryOfStructMethod
+                    : ActivatorCache.FactoryOfClassMethod;
+
+                Factory = CreateDelegate<InstanceFactory<T>>(factoryMethod.MakeGenericMethod(Type));
```

Comment:
Note that `MakeGenericMethod` is not AOT friendly operation. It does not have that different characteristics from just using Reflection.Emit.

Reply:
Don't know why I stuck with delegates if it's possible to write code without them as a single method.